### DISCUSSION TITLE
MSSQL mapping of data type JSON to NVARCHAR(max) is wrong

### DISCUSSION THREAD

**Discussion by leo-schick** [sqlalchemy/sqlalchemy#9355] [Usage Questions] [answered]
### Describe the bug

The data type `JSON` is mapped in dialect `mssql` to `NVARCHAR(max)`, see [here](https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/dialects/mssql/json.py#L19).

I created now a table on Azure Synapse Serverless via SQLAlchemy. This table is a external one reading from a parquet file. The creation of the table works fine (I use a custom script to generate the ddl `CREATE EXTERNAL TABLE` sql statement but I use the semantics of sqlalchemy to generate the column list).

But when I read it, I get the error shown below.

It works when I use `VARCHAR(max)` instead of `NVARCHAR(max)` as datatype. I see this as a bug.

I see the following solutions to fix this

1. Change the general mapping to VARCHAR(max)
2. Use a custom mapping when the destination server is a Azure Synapse instance.

### SQLAlchemy Version in Use

1.4.46 but it looks to be still present in the main branch when I look into the source code

### DBAPI (i.e. the database driver)

mssql

### Python Version

3.9

### Operating system

Linux

### To Reproduce

```python
from sqlalchemy import MetaData, Table, JSON, Column

table = Table('advertiser_transactions', MetaData(),
              Column('commissionAmount', JSON))

eng          = ...         # You have to initialize manually the engine here; not part of this code
storage_name = 'DATALAKE'  # define your created storage
format_name  = 'PARQUET'   # define your parquet file format name in Azure Synapse

dialect = eng.dialect

options = {
    'LOCATION': '"my_path"',
    'DATA_SOURCE': dialect.identifier_preparer.quote_identifier(storage_name),
    'FILE_FORMAT': dialect.identifier_preparer.quote_identifier(format_name),
}

columns = []
for column in table.columns:
    col = dialect.identifier_preparer.format_column(column)
    col += ' ' + dialect.type_compiler.process(column_type, type_expression=column)
    columns.append(col)

sql_statement = 'CREATE EXTERNAL TABLE '
            + dialect.identifier_preparer.format_table(table)
            + '\n(\n\t'
            + ',\n\t'.join(columns)
            + '\n)\n'
            + 'WITH (\n\t'
            + ',\n\t'.join([f'{k} = {v}' for k, v in options.items()])
            + '\n);'

print(sql_statement)
```

### Error

```
Statement ID: {320FC0F4-41C6-4EFA-AA73-DED761935F69} | Query hash: 0x9BE06AF1CF3F05D | Distributed request ID: {CB34143A-3CAC-4955-AE08-94ED47DEBE06}. Total size of data scanned is 1 megabytes, total size of data moved is 0 megabytes, total size of data written is 0 megabytes.
Msg 15814, Level 16, State 1, Line 2
Column 'commissionAmount' of type 'NVARCHAR' is not compatible with external data type 'JSON string. (underlying parquet nested/repeatable column must be read as VARCHAR or CHAR)'. File/External table name: 'advertiser_transactions'.

Completion time: 2023-02-23T10:41:56.0536271+01:00
```

**Comment by CaselIT** [comment, marked as answer]
Hi,

Since your use case is quite specific I don't think changing the JSON type in mssql is the best approach here.

I think you can probably use with_variant https://docs.sqlalchemy.org/en/20/core/type_api.html#sqlalchemy.types.TypeEngine.with_variant to configure a custom type in mssql. Or alternatively just use varchar in the table defintion.

Since your use case is quite specific I don't think changing the JSON type in mssql is something that can be generalized

**Reply by CaselIT** [reply]
cc @gordthompson

**Reply by gordthompson** [reply]
Like @CaselIT says, you can just declare the column as VARCHAR:

```python
import sqlalchemy as sa

engine = sa.create_engine("mssql+pyodbc://scott:tiger^5HHH@mssql_199")

tbl = sa.Table(
    "advertiser_transactions",
    sa.MetaData(),
    sa.Column("commissionAmount", sa.VARCHAR),
)
tbl.drop(engine, checkfirst=True)
engine.echo = True
tbl.create(engine)
"""
CREATE TABLE advertiser_transactions (
	[commissionAmount] VARCHAR(max) NULL
)
"""
```

**Reply by leo-schick** [reply]
This is true, but that is just a workaround. It doesn't solve the problem. I don't get why JSON should be written into NVARCHAR(max) which is twice the size of a VARCHAR(MAX).

**Reply by leo-schick** [reply]
Maybe the example is not the best to point this out.

**Reply by CaselIT** [reply]
> This is true, but that is just a workaround. It doesn't solve the problem. I don't get why JSON should be written into NVARCHAR(max) which is twice the size of a VARCHAR(MAX).

Because json can contain any unicode character and varchar in general does not allow them

Again your use case is very limited, so it's unreasonable to expect a change for everyone just to satisfy your use case, considering that the change may not work in all cases

**Reply by leo-schick** [reply]
@CaselIT Well, ok. Then it might be the right way to build a own custom dialect for Azure Synapse, right?

**Reply by CaselIT** [reply]
No need to do that. You can just customize how json is compiled with https://docs.sqlalchemy.org/en/20/core/compiler.html#changing-compilation-of-types

**Reply by zzzeek** [reply]
> This is true, but that is just a workaround. It doesn't solve the problem. I don't get why JSON should be written into NVARCHAR(max) which is twice the size of a VARCHAR(MAX).

NVARCHAR allows Unicode.   VARCHAR does not.  it makes no sense to use VARCHAR for JSON if it can't store Unicode, refer to MSFT documentation at https://learn.microsoft.com/en-us/sql/relational-databases/json/json-data-sql-server?view=sql-server-ver16 where all the examples illustrate JSON data represented as NVARCHAR(max). 

>  Then it might be the right way to build a own custom dialect for Azure Synapse, right?
 
IIUC this is a specific feature of Azure Synapse Analytics called [Serverless SQL Pool](https://learn.microsoft.com/en-us/azure/synapse-analytics/sql/on-demand-workspace-overview) with a specific configuration element called "PARQUET" that itself does not seem to support the NVARCHAR datatype in some way.     There is no indication as yet that Azure Synapse overall does not support JSON functions with NVARCHAR data.   The doc at https://learn.microsoft.com/en-us/sql/relational-databases/json/json-data-sql-server?view=sql-server-ver16 states:

> JSON text is stored in VARCHAR or NVARCHAR columns and is indexed as plain text. Any SQL Server feature or component that supports text supports JSON, so there are almost no constraints on interaction between JSON and other SQL Server features. 

and that document itself states that it applies to Azure Synapse Analytics.    So the error message you are seeing would appear to contradict that documentation.      Also, can we confirm this is a modern MSFT product that does not support Unicode in JSON fields?    That seems very unusual ?

**Reply by leo-schick** [reply]
OK. I think we can close this discussion. Just to mention:

* I thought maybe the PARQUET format enforces the JSON to escape unicode and therefore Synapse requires VARCHAR. [But this is not the case.](https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#embedded-types)
* Using the *Change compilation of types* feature does not work for me. I need to support multiple database connections to multiple sql server "dialects"
* My idea to build a custom SQLAlchemy dialect for Synapse comes from that there are multiple data type issues compared to other MSSQL dialects. E.g. you can't use `VARCHAR(max)`. You have to write `VARCHAR(8000)`. I was planning to build it on the existing mssql dialect - the same way as the databricks sqlalchemy dialect is based on the spark sqlalchemy dialect.

**Reply by zzzeek** [reply]
if you did custom compilation for types, you get the SQL Server `Dialect` object.  you can look at the version string and determine what kind of SQL Server database it is.

**Comment by zzzeek** [comment]
I dont understand the context for the error .  Can you provide the complete stack trace please?

**Reply by leo-schick** [reply]
The error comes when you try to read the external table. Sqlalchemy does not throw any error, so, there is no stacktrace
